src/Simplex/Chat/Store.hs

```diff
@@ -380,81 +394,99 @@ getGroup st User {userId, userContactId} localDisplayName =
-      DB.execute
-        db
-        [sql|
-          INSERT INTO connections
-            (user_id, agent_conn_id, conn_status, conn_type, group_member_id) VALUES (?,?,?,?,?);
-        |]
-        (userId, agentConnId, ConnNew, ConnMember, groupMemberId)
+toGroupMember :: Int64 -> GroupMemberRow -> GroupMember
+toGroupMember userContactId (groupMemberId, memberId, memberRole, memberStatus, invitedById, memberContactId, displayName, fullName) =
+  let memberProfile = Profile {displayName, fullName}
+      invitedBy = toInvitedBy userContactId invitedById
+   in GroupMember {groupMemberId, memberId, memberRole, memberStatus, invitedBy, memberProfile, memberContactId}
+
+createGroupMember :: StoreMonad m => SQLiteStore -> TVar ChaChaDRG -> User -> Int64 -> Contact -> GroupMemberRole -> ConnId -> m GroupMember
+createGroupMember st gVar user groupId contact memberRole agentConnId =
+  liftIOEither . withTransaction st $ \db ->
+    createWithRandomId gVar $ \memId -> do
+      member <- createContactMember_ db user groupId contact (memId, memberRole) GSMemInvited IBUser
+      groupMemberId <- insertedRowId db
+      createMemberConnection_ db (userId user) groupMemberId agentConnId
+      pure member
+
+createMemberConnection :: MonadUnliftIO m => SQLiteStore -> UserId -> Int64 -> ConnId -> m ()
+createMemberConnection st userId groupMemberId agentConnId =
+  liftIO . withTransaction st $ \db -> createMemberConnection_ db userId groupMemberId agentConnId
 
-createMember_ :: DB.Connection -> Int64 -> Int64 -> GroupMemberRole -> GroupMemberStatus -> Maybe Int64 -> ByteString -> IO ()
-createMember_ db groupId contactId memberRole memberStatus invitedBy memberId =
-  DB.executeNamed
+updateGroupMemberStatus :: MonadUnliftIO m => SQLiteStore -> UserId -> Int64 -> GroupMemberStatus -> m ()
+updateGroupMemberStatus _st _userId _groupMemberId _memberStatus = pure ()
+
+createMemberConnection_ :: DB.Connection -> UserId -> Int64 -> ConnId -> IO ()
+createMemberConnection_ db userId groupMemberId agentConnId =
+  DB.execute
     db
     [sql|
-      INSERT INTO group_members
-        ( group_id, member_id, member_role, member_status, invited_by,
-          contact_profile_id, contact_id)
-      VALUES
-        (:group_id,:member_id,:member_role,:member_status,:invited_by,
-          (SELECT contact_profile_id FROM contacts WHERE contact_id = :contact_id),
-          :contact_id)
+      INSERT INTO connections
+        (user_id, agent_conn_id, conn_status, conn_type, group_member_id) VALUES (?,?,?,?,?);
     |]
-    [ ":group_id" := groupId,
-      ":member_id" := memberId,
-      ":member_role" := memberRole,
-      ":member_status" := memberStatus,
-      ":invited_by" := invitedBy,
-      ":contact_id" := contactId
-    ]
+    (userId, agentConnId, ConnNew, ConnMember, groupMemberId)
+
+createContactMember_ :: IsContact a => DB.Connection -> User -> Int64 -> a -> MemberInfo -> GroupMemberStatus -> InvitedBy -> IO GroupMember
```

Comment:
nitpick, but please move this right after `createGroupMember`? it's strangely separated from the last non utility method that uses it

Reply:
the idea was to have all non-exported methods in the bottom together

Reply:
okay